jobs/batch_jobs/dashboard_stats_one_off.py

```diff
@@ -0,0 +1,113 @@
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS-IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""One-off job for populating weekly dashboard stats."""
+
+from __future__ import absolute_import  # pylint: disable=import-only-modules
+from __future__ import unicode_literals  # pylint: disable=import-only-modules
+
+from core.domain import user_services
+from core.platform import models
+from jobs import base_jobs
+from jobs import job_utils
+from jobs.io import ndb_io
+from jobs.types import job_run_result
+import feconf
+
+import apache_beam as beam
+
+datastore_services = models.Registry.import_datastore_services()
+(user_models,) = models.Registry.import_models([models.NAMES.user])
+
+
+class DashboardStatsOneOffJob(base_jobs.JobBase):
+    """One-off job for populating weekly dashboard stats for all registered
+    users who have a non-None value of UserStatsModel.
+    """
+
+    def run(self):
+        user_settings_models = (
+            self.pipeline
+            | 'Get all UserSettingsModel' >> ndb_io.GetModels(
+                user_models.UserSettingsModel.query(), self.datastoreio_stub)
+        )
+
+        user_stats_models = (
+            self.pipeline
+            | 'Get all UserStatsModel' >> ndb_io.GetModels(
+                user_models.UserStatsModel.query(), self.datastoreio_stub)
+        )
+
+        # Creates UserStatsModels if it does not exists.
+        created_models = (
+            (user_settings_models, user_stats_models)
+            # Combines the two pcolls
+            | 'Merge PCollections' >> beam.Flatten()
+            # Returns a PCollection of
+            # Tuple[model.id, Tuple[user_settings_models, user_stats_models]].
+            | beam.GroupBy(lambda m: m.id)
+            # Discards model.id from the PCollection.
+            | beam.Values()
+            # Only keep groupings that indicate that the UserStatsModel is missing.
+            | beam.Filter(
+                lambda models: len(models) == 1 and
+                job_utils.get_model_kind(models[0]) == 'UserSettingsModel')
+            # Creates the missing UserStatsModels.
+            | beam.ParDo(CreateUserStatsModel())
+        )
+        
+        (
+            (user_stats_models, created_models)
+            | beam.Flatten()
+            | beam.ParDo(UpdateDashboardStatsLog())
+            | ndb_io.PutModels(self.datastoreio_stub)
+        )
+
+        job_result = (
+            self.pipeline
+            | beam.Create([job_run_result.JobRunResult(stdout='SUCCESS')])
+        )
+
+        return job_result
+
+
+class CreateUserStatsModel(beam.DoFn):
+
+    def process(self, user_settings_model):
```

Comment:
This type is actually `user_settings_models: list[UserSettingsModel]`.

Reply:
I had handled it by choosing the first element, but I have modified it but still the backend test seems to be frozen and there is no response.

Reply:
Yep, I've found the same thing while working on a different job. I think we'll just have to wait until the Python 3 PR has been merged, I wasn't able to figure out why the PutModels causes the pipeline to freeze.

Reply:
Ok, Thanks.